dkg-primitives/src/rounds/offline.rs

```diff
@@ -0,0 +1,305 @@
+			_ => Ok(true),
+		}
+	}
+
+	fn get_outgoing(&mut self) -> Vec<DKGOfflineMessage> {
+		match self {
+			Self::Started(offline_rounds) => offline_rounds.get_outgoing(),
+			_ => vec![],
+		}
+	}
+
+	fn handle_incoming(&mut self, data: DKGOfflineMessage, at: C) -> Result<(), DKGError> {
+		match self {
+			Self::NotStarted(pre_offline_rounds) => pre_offline_rounds.handle_incoming(data, at),
+			Self::Started(offline_rounds) => offline_rounds.handle_incoming(data, at),
+			_ => Ok(()),
+		}
+	}
+
+	fn is_finished(&self) -> bool {
+		match self {
+			Self::Started(offline_rounds) => offline_rounds.is_finished(),
+			_ => true,
+		}
+	}
+
+	fn try_finish(self) -> Result<Self, DKGError> {
+		match self {
+			Self::Started(ref offline_rounds) =>
+				if offline_rounds.is_finished() {
+					Ok(self)
+				} else {
+					Err(DKGError::SMNotFinished)
+				},
+			_ => Ok(self),
+		}
+	}
+}
+
+/// Pre-offline rounds
+
+pub struct PreOfflineRounds {
+	signer_set_id: SignerSetId,
+	pub pending_offline_msgs: Vec<DKGOfflineMessage>,
+}
+
+impl PreOfflineRounds {
+	pub fn new(signer_set_id: SignerSetId) -> Self {
+		Self { signer_set_id, pending_offline_msgs: Vec::default() }
+	}
+}
+
+impl<C> DKGRoundsSM<DKGOfflineMessage, Vec<DKGOfflineMessage>, C> for PreOfflineRounds
+where
+	C: AtLeast32BitUnsigned + Copy,
+{
+	fn handle_incoming(&mut self, data: DKGOfflineMessage, _at: C) -> Result<(), DKGError> {
+		self.pending_offline_msgs.push(data);
+		Ok(())
+	}
+
+	fn is_finished(&self) -> bool {
+		true
+	}
+
+	fn try_finish(self) -> Result<Vec<DKGOfflineMessage>, DKGError> {
+		Ok(self.pending_offline_msgs)
+	}
+}
+
+/// Offline rounds
+
+pub struct OfflineRounds<Clock>
+where
+	Clock: AtLeast32BitUnsigned + Copy,
+{
+	params: SignParams,
+	started_at: Clock,
+	round_key: Vec<u8>,
+	offline_stage: OfflineStage,
+}
+
+impl<C> OfflineRounds<C>
+where
+	C: AtLeast32BitUnsigned + Copy,
+{
+	pub fn new(
+		params: SignParams,
+		started_at: C,
+		round_key: Vec<u8>,
+		offline_stage: OfflineStage,
+	) -> Self {
+		Self { params, started_at, round_key, offline_stage }
+	}
+}
+
+impl<C> DKGRoundsSM<DKGOfflineMessage, CompletedOfflineStage, C> for OfflineRounds<C>
+where
+	C: AtLeast32BitUnsigned + Copy,
+{
+	/// Proceed to next step
+
+	fn proceed(&mut self, at: C) -> Result<bool, DKGError> {
+		trace!(target: "dkg", "🕸️  OfflineStage party {} enter proceed", self.params.party_index);
+
+		let offline_stage = &mut self.offline_stage;
+
+		if offline_stage.wants_to_proceed() {
+			info!(target: "dkg", "🕸️  OfflineStage party {} wants to proceed", offline_stage.party_ind());
+			trace!(target: "dkg", "🕸️  before: {:?}", offline_stage);
+			// TODO, handle asynchronously
+			match offline_stage.proceed() {
+				Ok(_) => {
+					trace!(target: "dkg", "🕸️  after: {:?}", offline_stage);
+				},
+				Err(err) => {
+					println!("Error proceeding offline stage {:?}", err);
+					match err {
+						gg20_sign::Error::ProceedRound(proceed_err) => match proceed_err {
```

Comment:
You can do the following here instead of matching on every single one of them you could match on all of them if they are all have the same inner struct.

```rust
match proceed_err {
  gg20_sign::rounds::Error::Round1(e)
  | gg20_sign::rounds::Error::Round2Stage4(e)
  | gg20_sign::rounds::Round3::Round3(e)
  | // and so on here
  => return Err(DKGError::OfflineMisbehaviour { bad_actors: vec_usize_to_u16(e.bad_actors), }),
}
```